src/app/MessageDef.h

```diff
@@ -0,0 +1,816 @@
+ */
+class ListBuilderBase : public BuilderBase
+{
+protected:
+    ListBuilderBase(void);
+
+public:
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter, const uint8_t aContextTagToUse);
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter);
+};
+
+/**
+*  @brief
+*    CHIP interaction model Event Path definition
+*
+*/
+namespace AttributePath {
+    enum
+    {
+        kCsTag_EndpointId      = 0,
+        kCsTag_ClusterId       = 1,
+        kCsTag_FieldTag        = 2,
+        kCsTag_ListIndex       = 3,
+    };
+
+    class Parser;
+    class Builder;
+}; // namespace AttributePath
+
+/**
+*  @class Parser
+*
+*  @brief
+*    Parses a Attribute Path container
+*/
+
+class AttributePath::Parser : public ParserBase
+{
+public:
+    // aReader has to be on the element of EventPath container
+    CHIP_ERROR Init(const chip::TLV::TLVReader & aReader);
+
+    // Roughly verify the schema is right, including
+    // 1) all mandatory tags are present
+    // 2) no unknown tags
+    // 3) all elements have expected data type
+    // 4) any tag can only appear once
+    CHIP_ERROR CheckSchemaValidity(void) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetEndpointId(uint32_t * const apEndpointId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetClusterId(uint32_t * const apClusterId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetFieldTag(uint8_t * const apEventId) const;
+
+    // CHIP_END_OF_TLV if there is no such element
+    // CHIP_ERROR_WRONG_TLV_TYPE if there is such element but it's not any of the defined unsigned integer types
+    CHIP_ERROR GetListIndex(uint8_t * const apListIndex) const;
+
+    // Get a TLVReader at the additional tags section. Next() must be called before accessing it.
+    CHIP_ERROR GetTags(chip::TLV::TLVReader * const apReader) const;
+};
+
+class AttributePath::Builder : public BuilderBase
+{
+public:
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter);
+
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter, const uint8_t aContextTagToUse);
+
+    AttributePath::Builder & EndpointId(const uint32_t aEndpointId);
+    AttributePath::Builder & ClusterId(const uint32_t aClusterId);
+    AttributePath::Builder & FieldTag(const uint8_t aFieldTag);
+    AttributePath::Builder & ListIndex(const uint16_t aEndpointId);
+
+    AttributePath::Builder & EndOfAttributePath(void);
+
+private:
+    CHIP_ERROR _Init(chip::TLV::TLVWriter * const apWriter, const uint64_t aTagInApiForm);
+};
+
+/**
+*  @brief
+*    Chip Event Path List definition
+*/
+namespace AttributePathList {
```

Comment:
I would suggest not closing the namespace right away, and then we do not need to forward-declare parser.

likely applies to the other namespaces within this file.

Reply:
Done, thanks